src/gui/SampleThumbnail.cpp

```diff
@@ -0,0 +1,216 @@
+/*
+ * SampleThumbnail.cpp
+ *
+ * Copyright (c) Copyright (c) 2024 Khoi Dau <casboi86@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "SampleThumbnail.h"
+
+namespace lmms {
+
+/* DEPRECATED; functionality is kept for testing conveniences */
+bool SampleThumbnail::selectFromGlobalThumbnailMap(const Sample& inputSample)
+{
+	const auto samplePtr = inputSample.buffer();
+	const auto name = inputSample.sampleFile();
+	const auto end = s_sampleThumbnailCacheMap.end();
+
+	if (const auto list = s_sampleThumbnailCacheMap.find(name); list != end)
+	{
+		m_thumbnailCache = list->second;
+		return true;
+	}
+
+	m_thumbnailCache = std::make_shared<ThumbnailCache>();
+	s_sampleThumbnailCacheMap.insert(std::make_pair(name, m_thumbnailCache));
+	return false;
+}
+
+/* DEPRECATED; functionality is kept for testing conveniences */
+void SampleThumbnail::cleanUpGlobalThumbnailMap()
+{
+	auto map = s_sampleThumbnailCacheMap.begin();
+	while (map != s_sampleThumbnailCacheMap.end())
+	{
+		// All sample thumbnails are destroyed, a.k.a sample goes out of use
+		if (map->second.use_count() == 1)
+		{
+			s_sampleThumbnailCacheMap.erase(map);
+			map = s_sampleThumbnailCacheMap.begin();
+			continue;
+		}
+
+		map++;
+	}
+}
+
+SampleThumbnail::Thumbnail SampleThumbnail::generate(const size_t thumbnailSize, const SampleFrame* buffer, const size_t size)
+{
+	const auto sampleChunk = (size + thumbnailSize) / thumbnailSize;
+	auto thumbnail = Thumbnail(thumbnailSize);
+
+	for (auto tIndex = size_t{0}; tIndex < thumbnailSize; tIndex++)
+	{
+		const auto
+		sampleIndex = tIndex * size / thumbnailSize,
+		sampleChunkBound = std::min(sampleIndex + sampleChunk, size);
+
+		thumbnail[tIndex] = std::accumulate(buffer + sampleIndex, buffer + sampleChunkBound, Bit{});
+	}
+
+	return thumbnail;
+}
+
+SampleThumbnail::SampleThumbnail(const Sample& inputSample)
```

Comment:
Same thing: no need to pass the entire `Sample` object as we are only concerned of the audio buffer.